### ISSUE TITLE
Improving compatibility with browserify

### ISSUE THREAD

**Issue by mathisonian** [vega/vega#369]
I've been doing some work to improve the way that users can use vega and vega-lite through browserify (and webpack).

In particular the goal is to make it so that `npm install vega` works with browserify without any shims or the additional installation steps outlined in https://github.com/vega/vega/wiki/Vega-and-Browserify. The issues with the current structure are causing confusing such as #359 (and were also responsible for some of the troubles in #325).

The strategy is to remove browserify shims for libraries which will work fine without any shim (including d3, topojson, etc). See my fork with these changes [here](https://github.com/mathisonian/vega). To get this to work entirely will require changes to the following modules:

* https://github.com/vega/vega
* https://github.com/vega/datalib
* https://github.com/vega/vega-lite

there is also an outstanding pull request on jasondavies/d3-cloud#72 (which vega depends on) which should allow the removal of that shim as well.

I have forks with all of these changes ready to go, but I wanted to open a discussion to make sure these were welcome changes before opening a bunch of PRs. These changes would immediately be used by [lightning-viz/lightning](/lightning-viz/lightning) and [ellisonbg/altair](/ellisonbg/altair). 

/cc @freeman-lab @ellisonbg

**Comment by arvind** [comment]
Hi @mathisonian, thanks for opening the issue. The browserify-shim configuration allows us to require external dependencies (e.g., D3, TopoJSON, etc.) without packaging them up as part of Vega itself. But, perhaps there's another way to achieve that without using browserify-shim, and the confusion associated with it? 

We had use [these recipes](https://github.com/thlorenz/browserify-shim/wiki/browserify-shim-recipes) as a template for the current setup of `package.json`. As I understand it, browserify simply needs to know how to resolve a request for an external dependency correctly, if it isn't part of the package. We're open to ideas for how to better support this!

**Comment by mathisonian** [comment]
@arvind generally browserify-shims are used when you can't load the module through npm for some reason and instead end up including it on the page through a `<script>` tag. Since d3 and topojson are included in vega's dependencies already, browserify will know to pull those modules from npm during the install step and a shim shouldn't be required.

**Comment by ellisonbg** [comment]
On Wed, Sep 23, 2015 at 9:42 AM, Arvind Satyanarayan <
notifications@github.com> wrote:

> Hi @mathisonian <https://github.com/mathisonian>, thanks for opening the
> issue. The browserify-shim configuration allows us to require external
> dependencies (e.g., D3, TopoJSON, etc.) without packaging them up as part
> of Vega itself. But, perhaps there's another way to achieve that without
> using browserify-shim, and the confusion associated with it?
>

In general, we try to be disciplined about having things actually depend on
their dependencies and then have a single browserify step at the top level
of deployment. That is working really well for us in very complex
combinations of packages.

If you need to browserify just vega, you could have a separate repo
vega-browserify that npm depends on vega and bundles everything up.

Also - why is it important to not actually depend on d3 and topojson?


> We had use these recipes
> <https://github.com/thlorenz/browserify-shim/wiki/browserify-shim-recipes>
> as a template for the current setup of package.json. As I understand it,
> browserify simply needs to know how to resolve a request for an external
> dependency correctly, if it isn't part of the package. We're open to ideas
> for how to better support this!
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/vega/vega/issues/369#issuecomment-142657874>.
>



-- 
Brian E. Granger
Associate Professor of Physics and Data Science
Cal Poly State University, San Luis Obispo
@ellisonbg on Twitter and GitHub
bgranger@calpoly.edu and ellisonbg@gmail.com

**Comment by arvind** [comment]
Yes, this is behaviour we're currently targeting -- requiring external dependencies be loaded via `<script>` tags -- and is primarily motivated by browser-based deployments of Vega (i.e., the built `vega.js` and `vega.min.js` files). There are a couple of advantages that I see with the current setup:

1. In lieu of a more granular build system that allows for custom builds, decoupling Vega from its external dependencies allows browser deployments to only require those that are needed. For example, the [Vega homepage](http://vega.github.io/vega/) does not use TopoJSON or the Word Cloud Layout. 

2. Patch improvements to dependencies can be immediately used by users, rather than necessitating a patch release of Vega that simply rebuilds against the new versions. 

I should note that we're actively considering moving over to the new D3 submodules. datalib has already done so, and the D3 submodules it relies on are packaged up as part of datalib itself.

**Comment by mathisonian** [comment]
Ah cool @arvind, that makes sense. 

With my suggested changes I believe you can still get the benefits of (2) with the use of the carat or tilde when listing dependencies (http://stackoverflow.com/questions/22343224/difference-between-tilde-and-caret-in-package-json) without needing any releases on vega's part.

As for (1), I've seen this handled in a couple of ways. The first is to include, e.g. word cloud, as an "optional dependency" in package.json and wrap the `require('d3.layout.wordcloud')`, in a try/catch. Alternatively you could just depend on the user to install that package specifically if they want to use it an print an error message if its not available. For example, this is how https://github.com/sequelize/sequelize is able to ship as an ORM without specifically depending on sqlite, postgres, mysql, etc. etc.).

**Comment by mathisonian** [comment]
And :+1: for d3 submodules. I've been using them as much as possible for new visualization projects, have been working great so far

**Comment by jheer** [comment]
Re #1: We already use carat in for our dependencies, which works well if you are using npm or making your own build. However, that doesn't remove the need to release new browserify bundles if they include the dependencies. Not a huge issue, but there it is.

Another issue is that some web apps will mix-and-match vis tools, for example using D3 in places and Vega in others. One wants to avoid redundant imports in that case, but not everyone is bundling their apps up with browserify. The heart of the question here is what should be the default build vs. what should require a custom build.

In any case, yes, we'd like to switch to d3 v4 modules once their coverage is sufficient to replace d3 v3 proper. However, it's not clear when that will be. We'll see!

**Comment by mathisonian** [comment]
Yeah, these are all valid concerns, and I wouldn't want to introduce the need for more releases. 

I think one possibility would be to run the shim conditionally, so that things get shimmed when `npm build` is run from the vega repo, but the shim doesn't get triggered when installed via npm. (it looks like this should be possible using the `shim-config` option https://github.com/thlorenz/browserify-shim/issues/158).

**Comment by arvind** [comment]
Thanks for the pointer to conditional shimming! That may indeed help alleviate some of the confusion, while giving us the flexibility for external dependencies across npm and browserify builds.

**Comment by mathisonian** [comment]
cool! let me know and I'm happy to take a stab at a PR for any of this

**Comment by ellisonbg** [comment]
We are excited to see where this goes.

Matthew, you are in NYC right?

On Wed, Sep 23, 2015 at 11:54 AM, Matthew Conlen <notifications@github.com>
wrote:

> cool! let me know and I'm happy to take a stab at a PR for any of this
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/vega/vega/issues/369#issuecomment-142695828>.
>



-- 
Brian E. Granger
Associate Professor of Physics and Data Science
Cal Poly State University, San Luis Obispo
@ellisonbg on Twitter and GitHub
bgranger@calpoly.edu and ellisonbg@gmail.com

**Comment by mathisonian** [comment]
me too. yeah, @ellisonbg I'm based in nyc

**Comment by ellisonbg** [comment]
I am in NYC all of October, would be great to meet and chat more about this
stuff - maybe when Jeremy is in town as well.

On Thu, Sep 24, 2015 at 10:16 AM, Matthew Conlen <notifications@github.com>
wrote:

> me too. yeah, @ellisonbg <https://github.com/ellisonbg> I'm based in nyc
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/vega/vega/issues/369#issuecomment-142942934>.
>



-- 
Brian E. Granger
Associate Professor of Physics and Data Science
Cal Poly State University, San Luis Obispo
@ellisonbg on Twitter and GitHub
bgranger@calpoly.edu and ellisonbg@gmail.com

**Comment by freeman-lab** [comment]
Would be great! Could do most times this week starting Wed and through the weekend.

**Comment by mathisonian** [comment]
sounds good to me. lets move this off thread

**Comment by larf311** [comment]
Its seems that the last hurdle to making this package work seemlessly with browserify would be to change the transforms from devDependencies to regular dependencies. This way, this package's dev dependecies don't have to leak into the dev dependencies of downstream libraries.

You can find an example in the browserify handbook in the bottom of this section on creating reusable components (see how brfs is a regular dependency):

https://github.com/substack/browserify-handbook#reusable-components
